design/improve_error_handling_todo_list.md

```diff
@@ -0,0 +1,218 @@
+* api.py
+
+  * init function - These two exceptions can be replaced with MissingParameterException
+
+    * EmptyParameterError Account
+
+      ```python
+      self._account = account
+      if not self._account:
+          raise RuntimeError("Account cannot be empty!")
+      ```
+
+    * EmptyParameterError Url
+
+      * ```python
+        self._account = account
+        if not self._account:
+            raise RuntimeError("Account cannot be empty!")
+        ```
+
+  * login function-
+
+    * MissingRequiredParameterException
+
+      ```python
+      if not login_id or not password:
+          # TODO: Use custom error
+          raise RuntimeError("Missing parameters in login invocation!")
+      ```
+
+  * authenticate function
+
+    * MissingRequiredParameterException
+
+      * ```python
+        if not self.login_id or not self.api_key:
+            # TODO: Use custom error
+            raise RuntimeError("Missing parameters in authentication invocation!")
+        ```
+
+  * rotate_other_api_key function
+
+    * UnrotatableResource - We can rotate only user or host api key
+
+      ```python
+      if resource.type not in ('user', 'host'):
+          raise Exception("Error: Invalid resource type")
+      ```
+
+* client.py
+
+  * init function
+
+    * CredentialsParseError
+
+      ```python
+      try:
+          credentials = CredentialsFromFile(DEFAULT_NETRC_FILE)
+          loaded_netrc = credentials.load(loaded_config['url'])
+      except netrc.NetrcParseError as netrc_error:
+          raise Exception("Error: netrc is in an invalid format. "
+                          f"Reason: {netrc_error}") from netrc_error
+      except Exception as exception:
+          # pylint: disable=line-too-long
+          raise RuntimeError("Unable to authenticate with Conjur. Please log in and try again.") from exception
+      ```
+
+      About the `except Exception`  there we need to check which exception can be thrown . load functino can throw NotLogedInException (written in these document)
+
+* credentials_from_file.py
+
+  * load function
+
+    * NotLogedInException
+
+      ```python
+      if netrc_obj.hosts == {}:
+          raise Exception("You are already logged out")
+      ```
+
+      ```python
+      if netrc_auth == "":
+          raise Exception("You are already logged out")
+      ```
+
+* http_wrapper.py
+
+  * enable_http_logging
+
+* Init_controller.py
+
+  * get_server_certificate function
+
+    * EmptyParameterError
+
+      ```python
+      if self.conjurrc_data.appliance_url == '':
+          # pylint: disable=raise-missing-from
+          raise RuntimeError("Error: URL is required")
+      ```
+
+    * InvalidURLError
+
+      ```python
+      if url.scheme != 'https':
+          raise RuntimeError(f"Error: undefined behavior. Reason: The Conjur URL format provided "
+                 f"'{self.conjurrc_data.appliance_url}' is not supported.")
+      ```
+
+    * UserRefusedException - User dont trust certificate
```

Comment:
`ConfirmationException`? 

"User does not trust certificate

Reply:
Changed